### PR TITLE
refactor(dropdown): added 'before' events for events that fire after transition

### PR DESCRIPTION
**Pull request by Elijbet**
**Related Issue:** #4689

## Summary

Provide devs with more control of the components through additional hooks. For public events that are emitted after a transition finishes (in this case, opening/closing of the dropdown), we are now adding events that fire before the transition kicks off as well.

### CODE REVIEW

File: src/components/dropdown/dropdown.tsx [hunk 1]
```diff
@@ -439,14 +448,39 @@ export class Dropdown implements InteractiveComponent {
...
     }
   };
 
+  onTransitionRun = (event: TransitionEvent): void => {
+    if (event.propertyName === this.activeTransitionProp) {
+      this.active
+        ? this.openCloseEventEmitter("beforeOpen")
+        : this.openCloseEventEmitter("beforeClose");
+    }
+  };
```
Comment (eriklharper) [inline src/components/dropdown/dropdown.tsx:466]: I would recommend renaming this to `transitionRunHandler` to be more explicit that it is a handler function as opposed to an event.

File: src/components/dropdown/dropdown.tsx [hunk 2]
```diff
@@ -439,14 +448,39 @@ export class Dropdown implements InteractiveComponent {
...
+      beforeOpen: () => this.calciteDropdownBeforeOpen.emit(payload),
+      open: () => this.calciteDropdownOpen.emit(payload),
+      beforeClose: () => this.calciteDropdownBeforeClose.emit(payload),
+      close: () => this.calciteDropdownClose.emit(payload)
+    };
+    (
+      emitComponentState[componentVisibilityState] ||
+      emitComponentState["The component state is unknown."]
+    )();
+  }
```
Comment (eriklharper) [inline src/components/dropdown/dropdown.tsx:482]: In a similar vein, this function would be more appropriately titled `emitOpenCloseEvent` since it emits events, but is not actually an `EventEmitter` object.

File: src/components/dropdown/dropdown.e2e.ts [hunk 3]
```diff
@@ -539,8 +540,9 @@ describe("calcite-dropdown", () => {
 
     it("control max items displayed", async () => {
       const maxItems = 7;
-      const page = await newE2EPage({
-        html: html`<calcite-dropdown max-items="${maxItems}">
+      const page = await newE2EPage();
+      await page.setContent(html`
+        <calcite-dropdown max-items="${maxItems}">
```
Comment (eriklharper) [inline src/components/dropdown/dropdown.e2e.ts:545]: This doesn't look related to the changes in this PR, but more a refactor of how the page content is set.  Can you revert this back to the way it was since the shorthand method is preferred?

File: src/components/dropdown/dropdown.e2e.ts [hunk 4]
```diff
@@ -1024,4 +1026,56 @@ describe("calcite-dropdown", () => {
...
+    await page.setContent(html`
+      <calcite-dropdown>
+        <calcite-button slot="dropdown-trigger">Open dropdown</calcite-button>
+        <calcite-dropdown-group id="group-1">
+          <calcite-dropdown-item id="item-1"> Dropdown Item Content </calcite-dropdown-item>
+          <calcite-dropdown-item id="item-2" active> Dropdown Item Content </calcite-dropdown-item>
+          <calcite-dropdown-item id="item-3"> Dropdown Item Content </calcite-dropdown-item>
+        </calcite-dropdown-group>
+      </calcite-dropdown>
+    `);
```
Comment (eriklharper) [inline src/components/dropdown/dropdown.e2e.ts:1041]: Similarly here, you can simplify this with `await newE2EPage({ html: <html-goes-here> })`.

Comment (Elijbet) [reply src/components/dropdown/dropdown.e2e.ts:1041]: I believe setting page through `page.setContent` was @jcfranco's preference, as he corrected it a couple of times. From what I understand as opposed to `await newE2EPage()`, `page.setContent()` should have `await page.waitForChanges()` built in. I might be wrong though.

Comment (eriklharper) [reply src/components/dropdown/dropdown.e2e.ts:1041]: Lol, I recall Franco doing the exact opposite on some of my PRs where he recommended setting it the way I described above.  I honestly don't have a preference either way, just was following what I know I've encountered on reviews in the past.

Comment (Elijbet) [reply src/components/dropdown/dropdown.e2e.ts:1041]: Ok :) Maybe we should convene at some point and pick one :).

Comment (Elijbet) [reply src/components/dropdown/dropdown.e2e.ts:1041]: I put it on the agenda for the next time to agree on a preferred method of setting the page content.

File: src/components/dropdown/dropdown.e2e.ts [hunk 5]
```diff
@@ -1024,4 +1026,56 @@ describe("calcite-dropdown", () => {
...
+    const element = await page.find(`calcite-dropdown`);
+    const group = await page.find(`calcite-dropdown >>> .${CSS.calciteDropdownContent}`);
+
+    expect(await group.isVisible()).toBe(false);
+
+    const calciteDropdownBeforeOpenEvent = page.waitForEvent("calciteDropdownBeforeOpen");
+    const calciteDropdownOpenEvent = page.waitForEvent("calciteDropdownOpen");
```
[the remark this answers is not in the document]
Comment (Elijbet) [reply src/components/dropdown/dropdown.e2e.ts:1048]: It might be the transition timing, but when I don't intentionally `waitForEvent`, I get an error that `expected event "calciteDropdownOpen" to have been called 1 times, but was called 0 time`.

**Review by anveshmekala — APPROVED** [review]
:rocket:

File: src/components/dropdown/dropdown.e2e.ts (hunk 4 above)
Comment (eriklharper) [reply src/components/dropdown/dropdown.e2e.ts:1040]: Fine to leave as-is for now, shouldn't block merging.

**Review by eriklharper — APPROVED** [review]
🎉
